mkosi/__init__.py

```diff
@@ -1910,11 +1928,22 @@ def build_uki(
         )
 
         if sign_expected_pcr:
-            cmd += [
-                "--pcr-private-key", context.config.secure_boot_key,
-                "--pcr-banks", "sha1,sha256",
-            ]
-            options += ["--ro-bind", context.config.secure_boot_key, context.config.secure_boot_key]
+            cmd += ["--pcr-banks", "sha1,sha256"]
+            if context.config.secure_boot_key:
+                cmd += ["--pcr-private-key", context.config.secure_boot_key]
+                if (
+                    context.config.secure_boot_key_source.type == KeySource.Type.engine and
+                    context.config.secure_boot_certificate
+                ):
+                    cmd += [
+                        "--signing-engine", context.config.secure_boot_key_source.source,
+                        "--pcr-public-key", context.config.secure_boot_certificate,
+                    ]
+                    options += [
+                        "--ro-bind", context.config.secure_boot_certificate, context.config.secure_boot_certificate,
+                    ]
```

Comment:
This part confuses me. What does the public key have to do with the signing engine?

Reply:
@bluca Ok I hadn't noticed that we interpret `--pcr-public-key` as a certificate if `--signing-engine` is specified.

I did notice in the ukify source code that you forgot to convert the certificate to a public key for storage in the `pcrpkey` section which expects a public key and not a certificate, quote from the docs:

>A .pcrpkey section with a public key in the PEM format matching the signature data in the the .pcrsig section.

Reply:
That should be handled by this bit?

https://github.com/systemd/systemd/blob/main/src/boot/measure.c#L861

Reply:
@bluca That's for systemd-measure, I'm talking about this bit:

```python
    pcrpkey = opts.pcrpkey
    if pcrpkey is None:
        if opts.pcr_public_keys and len(opts.pcr_public_keys) == 1:
            pcrpkey = opts.pcr_public_keys[0]
        elif opts.pcr_private_keys and len(opts.pcr_private_keys) == 1:
            from cryptography.hazmat.primitives import serialization
            privkey = serialization.load_pem_private_key(pathlib.Path(opts.pcr_private_keys[0]).read_bytes(), password=None)
            pcrpkey = privkey.public_key().public_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PublicFormat.SubjectPublicKeyInfo,
            )

    sections = [
        # name,      content,         measure?
        ('.osrel',   opts.os_release, True ),
        ('.cmdline', opts.cmdline,    True ),
        ('.dtb',     opts.devicetree, True ),
        ('.uname',   opts.uname,      True ),
        ('.splash',  opts.splash,     True ),
        ('.pcrpkey', pcrpkey,         True ),
        ('.initrd',  initrd,          True ),

        # linux shall be last to leave breathing room for decompression.
        # We'll add it later.
    ]
```

When using `--signing-engine`, `.pcrpkey` will now contain a certificate instead of a public key.

Reply:
Ah yes you are right, will fix that